### ISSUE TITLE
Mark Write7BitEncodedInt and Read7BitEncodedInt as public

### ISSUE THREAD

**Issue by vermorel** [dotnet/corefx#25950]
7-bits encoding are useful in many situations. Yet, the method [BinaryWriter.Write7BitEncodedInt(int)](https://msdn.microsoft.com/en-us/library/system.io.binarywriter.write7bitencodedint(v=vs.110).aspx) is `protected` (idem for [BinaryReader.Read7BitEncodedInt](https://msdn.microsoft.com/en-us/library/system.io.binaryreader.read7bitencodedint(v=vs.110).aspx)) are not public. As a result, codebases end-up implementing a variant of `DummyBinaryReader` just access through inheritance the two properties. It would be much nicer if those two methods could be made public directly.

**Comment by jnm2** [comment]
Yes!

**Comment by danmosemsft** [comment]
This need a bit more formal writeup as an api proposal.

**Comment by karelz** [comment]
Can you please submit [formal API proposal](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md)? (see the example there)
It would be nice to comment more on how often devs work around the problem by inheritting (e.g. search GH?)

**Comment by jjxtra** [comment]
I very often write my own int and long 7 bit encoder in any project that needs compressed data size. Would be nice to have public versions of int and long.

**Comment by danmosemsft** [comment]
@jixtra this is blocked on API proposal. Would you like to write one up, per the link from @karelz?

**Comment by danmosemsft** [comment]
@jjxtra rather.

**Comment by jjxtra** [comment]
Doing it now!

**Comment by jjxtra** [comment]
Do I submit a new issue with the write up here, or somewhere else?

**Comment by jjxtra** [comment]
Submitted as issue #31038 31038

**Comment by karelz** [comment]
I think we should have only 1 issue. @jjxtra can you please move your proposal here? Thanks!

**Comment by jjxtra** [comment]
**Summary**
Right now BinaryReader and BinaryWriter have a read/write 7 bit encoded int method that is protected. These methods are currently used to prefix strings with a byte count. Having public methods would eliminate a lot of duplicate code and unnecessary inheritance.

**Rationale**
For anyone not inheriting these classes, we must roll our own 7 bit encoding methods, or inherit these classes. In addition, larger integer types like long and BigInteger also require custom code, but more or less share the same algorithm.

The following types would benefit from a public 7 bit encoding interface on BinaryReader and BinaryWriter:

- System.Int32
- System.Int64
- System.Numerics.BigInteger

I propose the following public methods be added for BinaryReader:

```
public class BinaryReader
{
    public int Read7BitEncodedInt32(); // returns the decoded int
    public long Read7BitEncodedInt64(); // returns the decoded long
    public System.Numerics.BigInteger Read7BitEncodedBigInteger(); // returns the decoded big integer
}
```

I propose the following public methods be added for BinaryWriter:

```
public class BinaryWriter
{
    public int Write7BitEncodedInt32(int value); // returns number of bytes written
    public int Write7BitEncodedInt64(long value); // returns number of bytes written
    public int Write7BitEncodedBigInteger(System.Numerics.BigInteger); // returns number of bytes written
}
```

**Notes**
Big integer may not seem to make sense at first, but it can still be useful, especially in data sets where the smallest numbers are the most common, but then larger more uncommon numbers also exist. If Big Integer proves too much a head-ache, at least having int and long would be very helpful to many developers.

The protected Write7BitEncodedInt and Read7BitEncodedInt methods can stay un-altered in order to not break existing code that is using inheritance.

**Comment by glenn-slayden** [comment]
(File As: department of pointless optimizations)<br><hr>

Regarding **Write7BitEncodedInt**, if you have a way to find the most-significant bit (MSB), you can unroll the code. For simplicity of demonstration here, I'll just use an **ntdll** import (declaration at the bottom), but ideally you'd want to use a JIT intrinsic, deBruijn, lookup-table, or other bit-scanning technique to stay in managed code.

Shown here as extension methods, but obviously they can easily be deployed as instance methods in **BinaryWriter** / **BinaryReader**. These functions are 100% fully compatible with the existing versions and exhaustively/conclusively tested (*i.e.*, roundtrip tested for every/each/all 2³² `int` values).

**Write7BitEncodedInt**

    public static void Write7BitEncodedInt(this BinaryWriter bw, int i)
    {
        switch (RtlFindMostSignificantBit((uint)i) / 7)
        {
            case 0:
                bw.BaseStream.WriteByte((byte)i);
                break;
            case 1:
                bw.BaseStream.WriteByte((byte)(i | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 7));
                break;
            case 2:
                bw.BaseStream.WriteByte((byte)(i /***/ | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 07 | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 14));
                break;
            case 3:
                bw.BaseStream.WriteByte((byte)(i /***/ | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 07 | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 14 | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 21));
                break;
            case 4:
                bw.BaseStream.WriteByte((byte)(i /***/ | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 07 | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 14 | 0x80));
                bw.BaseStream.WriteByte((byte)(i >> 21 | 0x80));
                bw.BaseStream.WriteByte((byte)((uint)i >> 28));
                break;
        }
    }

**Read7BitEncodedInt** (for this one, just some simple cleanups)

    public static int Read7BitEncodedInt(this BinaryReader br)
    {
        sbyte b;
        int r = -7, v = 0;
        do
            v |= ((b = br.ReadSByte()) & 0x7F) << (r += 7);
        while (b < 0);
        return v;
    }

As noted above. Ideally, some kind of JIT intrinsic, if available, would be great instead of this. Although the target native code is ideal, the transition to unmanaged code is undesirable. 

    [DllImport("ntdll")]
    static extern int RtlFindMostSignificantBit(ulong ul);

    // bsr rdx, rcx  
    // mov eax,0FFFFFFFFh  
    // movzx ecx, dl  
    // cmovne eax,ecx  
    // ret
